src/borrow_tracker/tree_borrows/tree/tests.rs

```diff
@@ -0,0 +1,674 @@
+        }
+    }
+
+    impl LocStateProtPair {
+        #[rustfmt::skip]
+        /// Two states (by convention `self` is the source and `other` is the target)
+        /// are "distinguishable" if there exists a sequence of instructions
+        /// that causes UB in the target but not in the source.
+        /// This implementation simply explores the reachable space
+        /// by all sequences of `TestEvent`.
+        /// This function can be instanciated with `RetX` and `RetY`
+        /// among `!` or `()` to resp. forbid/allow `x`/`y` to lose their
+        /// protector.
+        fn distinguishable<RetX, RetY>(&self, other: &Self, xy_rel: RelPosXY) -> bool
+        where
+            RetX: Exhaustive + 'static,
+            RetY: Exhaustive + 'static,
+        {
+            if self == other { return false; }
+            let mut states = vec![(self.clone(), other.clone())];
+            let mut seen = FxHashSet::default();
+            while let Some(state) = states.pop() {
+                if !seen.insert(state.clone()) { continue; };
+                let (source, target) = state;
+                for evt in <TestEvent<RetX, RetY>>::exhaustive() {
+                    // Generate successor states through events (accesses and protector ends)
+                    let Ok(new_source) = source.perform_test_event(&evt, xy_rel) else { continue; };
+                    let Ok(new_target) = target.perform_test_event(&evt, xy_rel) else { return true; };
+                    if new_source == new_target { continue; }
+                    states.push((new_source, new_target));
+                }
+            }
+            false
+        }
+    }
+
+    #[test]
+    #[should_panic]
+    // This is why `Reserved -> Frozen` on foreign read for protected references
+    // prevents the insertion of spurious reads: the transition can cause UB in the target
+    // later down the line.
+    fn reserved_frozen_protected_distinguishable() {
+        let xy_rel = RelPosXY::MutuallyForeign;
+        let source = LocStateProtPair {
+            x: LocStateProt {
+                state: LocationState::new(Permission::new_frozen()).with_access(),
+                prot: true,
+            },
+            y: LocStateProt {
+                state: LocationState::new(Permission::new_reserved(false)),
+                prot: true,
+            },
+        };
+        let acc = TestAccess { ptr: PtrSelector::X, kind: AccessKind::Read };
+        let target = source.perform_test_access(&acc, xy_rel).unwrap();
+        assert!(source.y.state.permission.is_reserved(None));
+        assert!(target.y.state.permission.is_frozen());
+        assert!(!source.distinguishable::<(), ()>(&target, xy_rel))
+    }
+
+    #[derive(Clone, Debug)]
+    struct Pattern {
+        xy_rel: RelPosXY,
+        x_retag_first: bool,
+        x_initial_state: LocationState,
+        y_initial_state: LocationState,
+        y_protected: bool,
+        y_return: bool,
```

Comment:
`y_return` seems to be completely ignored except for printing?

Reply:
Oh yeah, everything that `y_return` was in charge of is now handled by `RetY` on `distinguishable`.